src/object_store/src/object/s3.rs

```diff
@@ -936,85 +979,91 @@ impl Stream for S3ObjectIter {
+    match inner.borrow_mut() {
+        ObjectErrorInner::S3 {
+            should_retry,
+            inner,
+        } => {
+            let sdk_err = inner
+                .as_ref()
+                .downcast_ref::<SdkError<E, aws_smithy_runtime_api::http::Response<SdkBody>>>()
+                .unwrap();
+            let err_should_retry = match sdk_err {
                 SdkError::DispatchFailure(e) => {
                     if e.is_timeout() {
                         tracing::warn!(target: "http_timeout_retry", "{e:?} occurs, retry S3 get_object request.");
-                        return true;
+                        true
+                    } else {
+                        false
                     }
                 }
-                SdkError::ServiceError(e) => match e.err().code() {
-                    None => {
-                        if self.retry_unknown_service_error {
-                            tracing::warn!(target: "unknown_service_error", "{e:?} occurs, retry S3 get_object request.");
-                            return true;
+                SdkError::ServiceError(e) => {
+                    let mut retry = match e.err().code() {
+                        None => {
+                            if config.s3.developer.object_store_retry_unknown_service_error
+                                || config.s3.retry_unknown_service_error
+                            {
+                                tracing::warn!(target: "unknown_service_error", "{e:?} occurs, retry S3 get_object request.");
+                                true
+                            } else {
+                                false
+                            }
                         }
-                    }
-                    Some(code) => {
-                        if self
-                            .retryable_service_error_codes
-                            .iter()
-                            .any(|s| s.as_str().eq_ignore_ascii_case(code))
-                        {
-                            tracing::warn!(target: "retryable_service_error", "{e:?} occurs, retry S3 get_object request.");
-                            return true;
+                        Some(code) => {
+                            if config
+                                .s3
+                                .developer
+                                .object_store_retryable_service_error_codes
+                                .iter()
+                                .any(|s| s.as_str().eq_ignore_ascii_case(code))
+                            {
+                                tracing::warn!(target: "retryable_service_error", "{e:?} occurs, retry S3 get_object request.");
+                                true
+                            } else {
+                                false
+                            }
                         }
+                    };
+
+                    if !retry && e.raw().status().is_client_error() {
```

Comment:
Minor: this is a new case not covered before. How do we know 4xx is always retrible? For example, 403 seems permanent to me.

Reply:
right, 4xx client error should not been retry, i will remove it.